manual/custom_codecs/extras/README.md

```diff
@@ -0,0 +1,417 @@
+        .register(optionalIntCodec);
+
+TypeToken<Optional<Integer>> optionalIntToken = new TypeToken<Optional<Integer>>() {};
+
+// schema: create table example (id int primary key, v int);
+PreparedStatement pst = session.prepare("insert into example (id, v) values (?, ?)");
+
+Optional<Integer> v1 = Optional.empty();
+session.execute(pst.bind()
+        .setInt("id", 1)
+        .set("v", v1, optionalIntToken));
+// v1 saved as NULL
+
+Optional<Integer> v2 = Optional.of(12);
+session.execute(pst.bind()
+        .setInt("id", 2)
+        .set("v", v2, optionalIntToken));
+// v2 saved as 12
+```
+
+This example is a bit more complex: `Optional<Integer>` is a generic type, so we can't use simple statements, because
+they select codecs according to the runtime types of their values (and we would lose the `<Integer>` part with type
+erasure).
+
+For the same reason, we need to give a type hint when setting "v", in the form of a [TypeToken]. Note that this is an
+anonymous inner class; we recommend storing these tokens as constants in a utility class, to avoid creating them too
+often.
+
+[OptionalCodec]: http://docs.datastax.com/en/drivers/java/3.0/com/datastax/driver/extras/codecs/jdk8/OptionalCodec.html
+[Optional]: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html
+[TypeToken]: http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/reflect/TypeToken.html
+
+#### Guava
+
+[OptionalCodec][OptionalCodec_guava] transforms another codec to wrap its Java type in a Guava
+[Optional][Optional_guava].
+
+
+```java
+import com.datastax.driver.extras.codecs.guava.OptionalCodec;
+import com.google.common.base.Optional;
+import com.google.common.reflect.TypeToken;
+
+// The built-in driver codec, that maps CQL's int to Java's Integer
+TypeCodec<Integer> intCodec = TypeCodec.cint();
+
+// A codec that maps CQL's int to Guava's Optional<Integer>
+OptionalCodec<Integer> optionalIntCodec = new OptionalCodec<Integer>(intCodec);
+
+cluster.getConfiguration().getCodecRegistry()
+        .register(optionalIntCodec);
+
+TypeToken<Optional<Integer>> optionalIntToken = new TypeToken<Optional<Integer>>() {
+};
+
+// schema: create table example (id int primary key, v int);
+PreparedStatement pst = session.prepare("insert into example (id, v) values (?, ?)");
+
+Optional<Integer> v1 = Optional.absent();
+session.execute(pst.bind()
+        .setInt("id", 1)
+        .set("v", v1, optionalIntToken));
+// v1 saved as NULL
+
+Optional<Integer> v2 = Optional.of(12);
+session.execute(pst.bind()
+        .setInt("id", 2)
+        .set("v", v2, optionalIntToken));
+// v2 saved as 12
+```
+
+See the JDK8 Optional section above for explanations about [TypeToken].
+
+[OptionalCodec_guava]: http://docs.datastax.com/en/drivers/java/3.0/com/datastax/driver/extras/codecs/guava/OptionalCodec.html
+[Optional_guava]: http://docs.guava-libraries.googlecode.com/git-history/v16.0.1/javadoc/com/google/common/base/Optional.html
+
+
+### Primitive arrays
```

Comment:
Actually, this package also contain an `ObjectArrayCodec` class – not sure though if it's worth mentioning.